src/layers/misc/dist_embedding.cu

```diff
@@ -0,0 +1,732 @@
+  constexpr size_t warp_size = 32;
+  for (size_t i = threadIdx.x; i < n; i += warp_size) {
+    dest[i] = src[i];
+  }
+  __syncwarp();
+  return dest;
+}
+
+/** See El::AbstractDistMatrix::ColOwner. */
+__device__ __forceinline__
+size_t distmat_index_owner(size_t global_index, size_t align, size_t stride) {
+  return (global_index + align) % stride;
+}
+
+/** See El::AbstractDistMatrix::GlobalCol. */
+__device__ __forceinline__
+size_t distmat_global_index(size_t local_index, size_t shift, size_t stride) {
+  return shift + local_index * stride;
+}
+
+/** See El::AbstractDistMatrix::LocalCol. */
+__device__ __forceinline__
+size_t distmat_local_index(size_t global_index, size_t rank, size_t align, size_t stride) {
+  auto shift = (stride + rank - align) % stride;
+  if (global_index > shift) {
+    return (global_index - shift - 1) / stride + 1;
+  }
+  else {
+    return 0;
+  }
+}
+
+template <typename Kernel, typename... Args>
+inline void launch_cuda_kernel(
+  const Kernel& kernel,
+  dim3 grid_dims,
+  dim3 block_dims,
+  size_t shared_mem,
+  cudaStream_t stream,
+  Args... args) {
+  void* arg_list[] = {
+    const_cast<void*>(reinterpret_cast<const void*>(&args))...
+  };
+  CHECK_CUDA(
+    cudaLaunchKernel(
+      reinterpret_cast<const void*>(&kernel),
+      grid_dims,
+      block_dims,
+      arg_list,
+      shared_mem,
+      stream));
+}
+
+#ifdef LBANN_HAS_NVSHMEM
+template <typename Kernel, typename... Args>
+inline void launch_nvshmem_collective_kernel(
+  const Kernel& kernel,
+  dim3 grid_dims,
+  dim3 block_dims,
+  size_t shared_mem,
+  cudaStream_t stream,
+  Args... args) {
+  if (grid_dims.x == 0) {
+    grid_dims.y = 0;
+    grid_dims.z = 0;
+  }
+  void* arg_list[] = {
+    const_cast<void*>(reinterpret_cast<const void*>(&args))...
+  };
+  auto status = nvshmemx_collective_launch(
+    reinterpret_cast<const void*>(&kernel),
+    grid_dims,
+    block_dims,
+    arg_list,
+    shared_mem,
+    stream);
+  if (status != 0) {
+    LBANN_ERROR(
+      "Failed to launch NVSHMEM collective kernel ",
+      "(error ",status,")");
+  }
+}
+#endif // LBANN_HAS_NVSHMEM
+
+} // namespace <anon>
+
+// =============================================
+// Life cycle and setup
+// =============================================
+
+template <typename TensorDataType, data_layout Layout, El::Device Device>
+dist_embedding_layer<TensorDataType,Layout,Device>::~dist_embedding_layer()
+{
+#ifdef LBANN_HAS_NVSHMEM
+  if (m_embeddings_buffer != nullptr) {
+    nvshmem_free(m_embeddings_buffer);
+  }
+  if (m_workspace_buffer != nullptr) {
+    nvshmem_free(m_workspace_buffer);
+  }
+  if (m_metadata_buffer != nullptr) {
+    nvshmem_free(m_metadata_buffer);
+  }
+#endif // LBANN_HAS_NVSHMEM
+}
+
+template <typename TensorDataType, data_layout Layout, El::Device Device>
+void dist_embedding_layer<TensorDataType,Layout,Device>::attach_embeddings_to_shmem_buffer() {
+#ifndef LBANN_HAS_NVSHMEM
```

Comment:
I don't care for this style of cluttering the implementation of nearly every functional part of the code with a preprocessor switch and runtime error. This layer *requires* NVSHMEM or OpenSHMEM. It should not be constructible if you don't have at least one of those, so the runtime errors should never even be reachable. 

I would prefer a more static approach to this -- only compile this class in at all if you have one of those. Then C++ code will never be able to compile code that statically depends on it, and "dynamic code" will never be able to successfully create it because it will not have been registered with the factory.